backend-tests/tests/test_reporting.py

```diff
@@ -0,0 +1,1225 @@
+# Copyright 2022 Northern.tech AS
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+import logging
+from urllib import request
+import pytest
+import time
+import uuid
+
+import requests
+
+from testutils.api.client import ApiClient
+from testutils.infra.container_manager.kubernetes_manager import isK8S
+import testutils.api.deviceauth as deviceauth
+import testutils.api.inventory as inventory
+import testutils.api.reporting as reporting
+import testutils.api.useradm as useradm
+
+from testutils.common import (
+    clean_mongo,
+    create_user,
+    create_org,
+    make_accepted_device,
+    mongo,
+    useExistingTenant,
+)
+
+
+def assert_device_attributes(dev, api_dev):
+    for attr in dev["attributes"]:
+        assert attr in api_dev["attributes"], (
+            "Missing inventory attribute: %s; device attributes: %s"
+            % (attr, api_dev["attributes"])
+        )
+
+
+def dict_to_inventoryattrs(d, scope="inventory"):
+    attr_list = []
+    for key, value in d.items():
+        attr = {"name": key, "value": value}
+        if scope is not None:
+            attr["scope"] = scope
+        attr_list.append(attr)
+
+    return attr_list
+
+
+def add_devices_to_tenant(tenant, dev_inventories):
+    try:
+        tenant.devices
+    except AttributeError:
+        tenant.devices = []
+
+    useradmm = ApiClient(useradm.URL_MGMT)
+    devauthd = ApiClient(deviceauth.URL_DEVICES)
+    devauthm = ApiClient(deviceauth.URL_MGMT)
+    invd = ApiClient(inventory.URL_DEV)
+
+    for inv in dev_inventories:
+        user = tenant.users[0]
+        utoken = useradmm.call(
+            "POST", useradm.URL_LOGIN, auth=(user.name, user.pwd)
+        ).text
+        assert utoken != ""
+
+        tenant.api_token = utoken
```

Comment:
I think this can be moved outside the loop.

Reply:
https://github.com/mendersoftware/integration/pull/1894/commits/6f1ca8e8f0e87a827a771bef7ea326f1085899f7